src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,172 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
                 properties: null,
                 SyntaxFacts.GetText(firstToken.Kind())));
         }
+
+        /// <summary>
+        /// <para>In general, statements are considered multiline if any of the following span more than one line:</para>
+        /// <list type="bullet">
+        /// <item><description>The part of the statement preceding the embedded statement</description></item>
+        /// <item><description>The embedded statement itself</description></item>
+        /// <item><description>The part of the statement following the embedded statement</description></item>
+        /// </list>
+        /// <para>The third condition is not checked for <c>else</c> clauses because they are only considered multiline
+        /// when their embedded statement is multiline.</para>
+        /// </summary>
+        private static bool IsConsideredMultiLine(SyntaxNode statement, SyntaxNode embeddedStatement)
+        {
+            // Early return if the entire statement fits on one line
+            if (FormattingRangeHelper.AreTwoTokensOnSameLine(statement.GetFirstToken(), statement.GetLastToken()))
+            {
+                // The entire statement fits on one line. Examples:
+                //
+                //   if (something) return;
+                //
+                //   while (true) something();
+                return false;
+            }
+
+            // Check the part of the statement preceding the embedded statement
+            if (!FormattingRangeHelper.AreTwoTokensOnSameLine(embeddedStatement.GetFirstToken(), embeddedStatement.GetLastToken()))
```

Comment:
Consider: You have the specification written above, which has 3 sections (the part above hte embedded statement, the embedded statement itself, and the part after the embedded statement).  However, this code doesn't follow that order or reference those spec comments.  Consider making the order hte same, and then just using the same comment (or referencing it).  It would more strongly connect the spec to the impl.

Reply:
➡️ Now updated (in order and comments corrected)